streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -272,15 +294,22 @@ public synchronized void onChange(final StreamThread thread,
             }
             if (newState == StreamThread.State.PARTITIONS_REVOKED ||
                 newState == StreamThread.State.ASSIGNING_PARTITIONS) {
-                setState(State.REBALANCING);
-            } else if (newState == StreamThread.State.RUNNING) {
+                setState(State.REBALANCING, true);
+            } else if (newState == StreamThread.State.RUNNING && state() != State.RUNNING) {
+                // one thread is running, check others
                 for (final StreamThread.State state : threadState.values()) {
                     if (state != StreamThread.State.RUNNING) {
                         return;
                     }
                 }
-                setState(State.RUNNING);
+                setState(State.RUNNING, true);
+            } else if (newState == StreamThread.State.DEAD) {
+                // one thread died, check if we have enough threads running
+                if (threadState.size() == 0) {
+                    setState(State.NOT_RUNNING, true);
```

Comment:
No it doesn't. I'm about to add state info to the global thread as well. Also if the global thread fails, for whatever reason, that should be fatal to the entire streams instance IMO. So the global thread is special and if it goes to DEAD, then Kafka streams will go to NOT_RUNNING. cc @dguy .